polytracker/include/taintdag/section.hpp

```diff
@@ -0,0 +1,182 @@
+  // NOTE(hbrodin): Error exit if it is not within the allocated part of the
+  // section
+  size_t offset(span_t::iterator it) const {
+    if (it < mem_.begin() || it >= mem_.end()) {
+      error_exit(
+          "Can't compute offset of iterator that is not within section.");
+    }
+    return std::distance(mem_.begin(), it);
+  }
+
+  // Returns the offset of p, computed from beginning of section
+  // NOTE(hbrodin): Error exit if it is not within the allocated part of the
+  // section
+  size_t offset(uint8_t const *p) const {
+    if (p < &*mem_.begin() || &*mem_.end() <= p)
+      error_exit("Can't compute offset of pointer that is not within section.");
+    return p - &*mem_.begin();
+  }
+
+  // The full range of memory assigned to this section
+  span_t mem_;
+
+private:
+  // Current write position in mem_
+  span_t::iterator write_it_;
+
+  // Want to be able to call const-methods such as size() and still lock to
+  // ensure any pending write is protected.
+  mutable std::mutex m_;
+};
+
+// Convenience base for allocation of fixed size entries
+template <typename T> struct FixedSizeAlloc : SectionBase {
+  static constexpr size_t align_of = alignof(T);
+
+  FixedSizeAlloc(SectionBase::span_t rng) : SectionBase{rng} {
+    if (reinterpret_cast<uintptr_t>(&*rng.begin()) % align_of != 0)
+      error_exit("FixedSizeAlloc requires memory to be aligned on align_of.");
+
+    if (rng.size() % entry_size() != 0) {
+      error_exit(
+          "FixedSizeAlloc requires memory to be a multiple of entry_size().");
+    }
+  }
+
+  constexpr size_t entry_size() const {
+    return sizeof(T);
+  } // TODO(hbrodin): Handle alignment/padding issues in a good way
+
+  // Helper type to ensure object construction can be done while holding a lock
+  // to ensure exclusive access.
+  struct ConstructCtx {
+    // WriteCtx from SectionBase provides accesss to the memory range and is
+    // responsible for locking the section.
+    SectionBase::WriteCtx ctx;
+
+    // Reference to the newly constructed object.
+    T &t;
+  };
+
+  // Construct an object of type T in the section assigned memory.
+  // If successfull a ConstructCtx is returned. Callers have exclusive
+  // access to the section for the life time of the ConstructCtx.
+  // N.b. count()/size() competes for the same lock.
+  template <typename... Args>
+  std::optional<ConstructCtx> construct(Args &&...args) {
+    return map(SectionBase::write(entry_size()), [&](auto &ctx) {
+      return ConstructCtx{.ctx = std::move(ctx),
+                          .t = *new (&*ctx.mem.begin())
+                                   T{std::forward<Args>(args)...}};
+    });
+  }
+
+  // Constructs n instances as a sequence of T.
+  // Each T is constructed by successive invocation of generator.
+  template <typename F>
+  std::optional<std::span<T const>> construct_range(
+      size_t n,
+      F &&generator) /* TODO(hbrodin): require(generator(uint8_t *)*/ {
+    // TODO(hbrodin): Check n > 0
```

Comment:
This would just be a sanity check. We don't expect anyone to invoke this with `n` being `0`. It can for sure be passed the value `0`, but it doesn't make any sense, hence the sanity check. The `size_t` type is unsigned and thus a negative value would become a large positive value. It would most likely fail du to the call to `SectionBase::write` having a `mem_size` larger than the capacity of the section.